### ISSUE TITLE
needless_borrows_for_generic_arg fires incorrectly on &mut variable inside closure

### ISSUE THREAD

**Issue by zancas** [rust-lang/rust-clippy#13170]
### Summary

cliippy incorrectly warns that a borrow is needless here:

```
cargo clippy
   Compiling build_utils v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/build_utils)
    Checking zingoconfig v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingoconfig)
    Checking zingo-netutils v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-netutils)
    Checking zingo-memo v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-memo)
    Checking zingo-status v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-status)
   Compiling darkside-tests v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/darkside-tests)
    Checking zingo-sync v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-sync)
warning: the borrowed expression implements the required traits
  --> zingo-memo/src/lib.rs:83:28
   |
83 |         CompactSize::write(&mut w, typecode as usize)?;
   |                            ^^^^^^ help: change this to: `w`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
   = note: `#[warn(clippy::needless_borrows_for_generic_args)]` on by default
```

When I rerun with --fix, I see (the expected) `borrow of moved value`

```


```

This seems superficially similar to #13162 but it causes a compile time error, not a runtime error.

### Lint Name

cargo clippy --fix

### Reproducer

I tried this code:

```rust
cargo clippy --fix
```

I saw this happen:

```
cargo clippy --fix
   Compiling build_utils v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/build_utils)
    Checking zingoconfig v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingoconfig)
    Checking zingo-netutils v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-netutils)
    Checking zingo-memo v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-memo)
    Checking zingo-status v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-status)
   Compiling darkside-tests v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/darkside-tests)
    Checking zingo-sync v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-sync)
    Checking zingo-testvectors v0.1.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingo-testvectors)
   Compiling zingolib v0.2.0 (/home/nattyb/src/zingolabs/zingolibs/spent_in/zingolib)
warning: failed to automatically apply fixes suggested by rustc to crate `zingo_memo`

after fixes were automatically applied the compiler reported errors within these files:

  * zingo-memo/src/lib.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
warning: variable does not need to be mutable
  --> zingo-memo/src/lib.rs:75:40
   |
75 |     Vector::write(writer, &receivers, |mut w, receiver| {
   |                                        ----^
   |                                        |
   |                                        help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default

error[E0382]: use of moved value: `w`
  --> zingo-memo/src/lib.rs:84:28
   |
75 |     Vector::write(writer, &receivers, |mut w, receiver| {
   |                                        ----- move occurs because `w` has type `&mut W`, which does not implement the `Copy` trait
...
83 |         CompactSize::write(w, typecode as usize)?;
   |                            - value moved here
84 |         CompactSize::write(w, data.len())?;
   |                            ^ value used here after move
   |
help: if `W` implemented `Clone`, you could clone the value
  --> zingo-memo/src/lib.rs:68:46
   |
68 | pub fn write_unified_address_to_raw_encoding<W: Write>(
   |                                              ^ consider constraining this type parameter with `Clone`
...
83 |         CompactSize::write(w, typecode as usize)?;
   |                            - you could clone this value
```

### Version

```text
rustc --version -v
rustc 1.80.0 (051478957 2024-07-21)
binary: rustc
commit-hash: 051478957371ee0084a7c0913941d2a8c4757bb9
commit-date: 2024-07-21
host: x86_64-unknown-linux-gnu
release: 1.80.0
LLVM version: 18.1.7
```

**Comment by zancas** [comment]
I compare the output of:
`cargo +1.79 clippy`

to:

`cargo +1.80 clippy`

```
cargo +1.80 clippy
warning: the borrowed expression implements the required traits
  --> zingo-memo/src/lib.rs:83:28
   |
83 |         CompactSize::write(&mut w, typecode as usize)?;
   |                            ^^^^^^ help: change this to: `w`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
   = note: `#[warn(clippy::needless_borrows_for_generic_args)]` on by default
```

This is a regession in clippy 1.80.

**Comment by T-O-R-U-S** [comment]
+1 for this issue. I think this affects all things that take in `AsMut<Type>` and `Type: Copy`, because `Type` and `&mut Type` satisfy this -- but then this causes problems when I want something to be mutated by the function, instead of copied.

```rs
let mut my_num = 1;

increment(&mut my_num); // clippy goes silly.
// suggest this instead.
increment(my_num);

println!("{my_num}"); // my_num is 1, when it should be 2! 
```

**Comment by Jarcho** [comment]
Should have been fixed by #12892.

**Comment by little-dude** [comment]
This is not fixed, as of `clippy 0.1.95 (873d4682c7 2026-01-25)`. I have something similar to the example @mahmoud-moursy gave:

```rust
use std::net::Ipv6Addr;

pub trait Encoder: Sized {
    fn encode<S>(&self, buf: S)
    where
        S: AsMut<[u8]> + AsRef<[u8]>;
}

impl Encoder for Ipv6Addr {
    fn encode<S>(&self, mut buf: S)
    where
        S: AsMut<[u8]> + AsRef<[u8]>,
    {
        let slice = buf.as_mut();
        slice.copy_from_slice(&self.octets());
    }
}

fn main() {
    let mut buf = [0u8; 16];
    Ipv6Addr::new(0xfe80, 0, 0, 0, 0, 0, 0, 1).encode(&mut buf); // Breaks if we just pass buf, since we mutate a copy
    assert_eq!(buf, [0xfe, 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]);
}
```
